### PR TITLE
Disable offloading for routes if server offloading disabled

### PR DESCRIPTION
**Pull request by bondolo**
If `noOffloadsStrategy()` is specified as the server execution strategy then routes for `PredicateRouter`, `InOrderRouter` and `gRPC` will be offloaded. Previously it was necessary to additionally request no offloading in order to offload the route. Routes can still be individually offloaded by specifying a strategy.

### CODE REVIEW

**Comment by bondolo** [comment]
This PR is now complete and ready for review.

**Comment by bondolo** [comment]
The current PR includes four failing tests for one particular configuration of non-offloading Server :

When using the predicate router with different service models (blocking/async, streaming/aggregate) there is a conflict between applying the offloads typical for a service model and adhering to the "no offloads" configuration of the server. If the route provides no explicit execution strategy should the service be offloaded or not?

For cases where an explicit strategy is specified, it will be combined with the service model strategy. Part of the problem is that we can't tell the difference between the user provided strategy and the combined service model and user strategy.

**Review by idelpivnitskiy — COMMENTED** [review]
We are getting closer. Thanks for addressing all previous comments.

File: servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcRouter.java [hunk 1]
```diff
@@ -100,12 +98,11 @@
  * implementation of a <a href="https://www.grpc.io">gRPC</a> method.
  */
 final class GrpcRouter {
-    private static final Logger LOGGER = LoggerFactory.getLogger(GrpcRouter.class);
```
Comment (idelpivnitskiy) [inline servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcRouter.java]: At some point, this logger produced a nice debugging output:
```
2021-11-02 15:17:37,124                           main [DEBUG] GrpcRouter                     - route strategy for path=/helloworld.Greeter/SayHello : ctx=DEFAULT_HTTP_EXECUTION_STRATEGY route=null use=DEFAULT_HTTP_EXECUTION_STRATEGY
```
Don't see it anymore. Consider adding it back. 

For `use=` consider either logging a result of the merge or just say something like: using a combination of both.

File: servicetalk-examples/grpc/execution-strategy/src/main/resources/log4j2.xml [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    </Console>
+  </Appenders>
+  <Loggers>
+    <!-- Prints server start and shutdown -->
+    <Logger name="io.servicetalk.http.netty.H2ServerParentConnectionContext" level="DEBUG"/>
+
+    <!-- Prints default subscriber errors-->
+    <Logger name="io.servicetalk.concurrent.api" level="DEBUG"/>
+    <!-- Prints execution strategy computation results -->
+    <Logger name="io.servicetalk.grpc.api" level="DEBUG"/>
```
[the remark this answers is not in the document]
Comment (idelpivnitskiy) [reply servicetalk-examples/grpc/execution-strategy/src/main/resources/log4j2.xml:28]: As I mentioned in another comment, don't see this logger producing any output.

File: servicetalk-examples/grpc/execution-strategy/src/main/java/io/servicetalk/examples/grpc/strategies/ExecutionStrategyServer.java [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+            // -> route offloaded to global executor
+            System.out.printf("\n%d : defaultServer\n", port);
+            ServerContext defaultServer = GrpcServers.forPort(port++)
+                    .listenAndAwait((Greeter.GreeterService)
+                            (ctx, request) -> getReplySingle(request, "default server"));
+            closeEverything.prepend(defaultServer);
+
+            // No offloads strategy specified for all routes by configuration of async server.
+            // -> no offloading, route executed on IoExecutor
+            System.out.printf("\n%d : asyncServer\n", port);
```
Comment (idelpivnitskiy) [inline servicetalk-examples/grpc/execution-strategy/src/main/java/io/servicetalk/examples/grpc/strategies/ExecutionStrategyServer.java:67]: Using consistent names for server and client output helps to relate events:
```patch
Index: servicetalk-examples/grpc/execution-strategy/src/main/java/io/servicetalk/examples/grpc/strategies/ExecutionStrategyServer.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/servicetalk-examples/grpc/execution-strategy/src/main/java/io/servicetalk/examples/grpc/strategies/ExecutionStrategyServer.java b/servicetalk-examples/grpc/execution-strategy/src/main/java/io/servicetalk/examples/grpc/strategies/ExecutionStrategyServer.java
--- a/servicetalk-examples/grpc/execution-strategy/src/main/java/io/servicetalk/examples/grpc/strategies/ExecutionStrategyServer.java	(revision 3a6fb44a796e8ad431ba6483c835032fb3308ca0)
+++ b/servicetalk-examples/grpc/execution-strategy/src/main/java/io/servicetalk/examples/grpc/strategies/ExecutionStrategyServer.java	(date 1636076447671)
@@ -29,6 +29,7 @@
 import io.servicetalk.transport.api.ServerContext;
 
 import io.grpc.examples.strategies.Greeter;
+import io.grpc.examples.strategies.Greeter.GreeterService;
 import io.grpc.examples.strategies.HelloReply;
 import io.grpc.examples.strategies.HelloRequest;
 
@@ -54,36 +55,35 @@
             // executor will be closed last, servers are prepended before executor.
             closeEverything.append(executor);
 
-            // default config
+            // Default server configuration.
             // -> route offloaded to global executor
-            System.out.printf("\n%d : defaultServer\n", port);
+            System.out.printf("\n%d : default server\n", port);
             ServerContext defaultServer = GrpcServers.forPort(port++)
-                    .listenAndAwait((Greeter.GreeterService)
+                    .listenAndAwait((GreeterService)
                             (ctx, request) -> getReplySingle(request, "default server"));
             closeEverything.prepend(defaultServer);
 
-            // No offloads strategy specified for all routes by configuration of async server.
+            // No offloads strategy specified on the server, async route does not override its strategy.
             // -> no offloading, route executed on IoExecutor
-            System.out.printf("\n%d : asyncServer\n", port);
+            System.out.printf("\n%d : no offloading server, async route\n", port);
             ServerContext asyncServer = GrpcServers.forPort(port++)
                     .initializeHttp(init -> init.executionStrategy(noOffloadsStrategy()))
-                    .listenAndAwait((Greeter.GreeterService)
-                            (ctx, request) -> getReplySingle(request, "no offloading server"));
+                    .listenAndAwait((GreeterService)
+                            (ctx, request) -> getReplySingle(request, "no offloading server, async route"));
             closeEverything.prepend(asyncServer);
 
-            // No offloads strategy specified for all routes by configuration of blocking server.
+            // No offloads strategy specified on the server, blocking route does not override its strategy.
             // -> no offloading, route executed on IoExecutor
-            System.out.printf("\n%d : blockingServer\n", port);
+            System.out.printf("\n%d : no offloading server, blocking route\n", port);
             ServerContext blockingServer = GrpcServers.forPort(port++)
                     .initializeHttp(init -> init.executionStrategy(noOffloadsStrategy()))
                     .listenAndAwait((Greeter.BlockingGreeterService)
-                            (ctx, request) -> getReply(request, "server blocking"));
+                            (ctx, request) -> getReply(request, "no offloading server, blocking route"));
             closeEverything.prepend(blockingServer);
 
-            // No offloads strategy specified for all routes by configuration of streaming server.
-            // Route attempts to use default strategy
+            // No offloads strategy specified on the server, route overrides it to use the default strategy.
             // -> route offloaded to global executor
-            System.out.printf("\n%d : noOffloadsServerRouteOffloads\n", port);
+            System.out.printf("\n%d : no offloading server, default offloading for the route\n", port);
             ServerContext noOffloadsServerRouteOffloads = GrpcServers.forPort(port++)
                     .initializeHttp(init -> init.executionStrategy(noOffloadsStrategy()))
                     .listenAndAwait(new Greeter.ServiceFactory.Builder()
@@ -93,10 +93,9 @@
                             .build());
             closeEverything.prepend(noOffloadsServerRouteOffloads);
 
-            // No offloads strategy specified for all routes by configuration of streaming server.
-            // Route uses custom strategy
+            // No offloads strategy specified on the server, route overrides it to use a custom strategy.
             // -> route offloaded to global executor
-            System.out.printf("\n%d: noOffloadsServerRouteOffloadCustom\n", port);
+            System.out.printf("\n%d: no offloading server, custom offloading for the route\n", port);
             ServerContext noOffloadsServerRouteOffloadCustom = GrpcServers.forPort(port++)
                     .initializeHttp(init -> init.executionStrategy(noOffloadsStrategy()))
                     .listenAndAwait(new Greeter.ServiceFactory.Builder().sayHello(CUSTOM_STRATEGY,
@@ -105,37 +104,37 @@
                             .build());
             closeEverything.prepend(noOffloadsServerRouteOffloadCustom);
 
-            // Server custom executor, routes are offloaded to executor
+            // Server with a default strategy but a custom executor, route does not override its strategy.
             // -> route offloaded to custom executor
-            System.out.printf("\n%d : customExecutorServer\n", port);
+            System.out.printf("\n%d : server with a default offloading and a custom executor\n", port);
             ServerContext customExecutorServer = GrpcServers.forPort(port++)
                     .initializeHttp(init -> init.executor(executor))
-                    .listenAndAwait((Greeter.GreeterService) (ctx, request) ->
+                    .listenAndAwait((GreeterService) (ctx, request) ->
                                     getReplySingle(request, "server with a default offloading and a custom executor"));
             closeEverything.prepend(customExecutorServer);
 
-            // Server has default configuration
-            // Route attempts to use no offloads strategy, which is ignored.
+            // Server has default configuration, route attempts to use no offloads strategy, which is ignored.
             // (Too late, already offloaded at the server level)
             // -> route offloaded to global executor
-            System.out.printf("\n%d : noOffloadsRoute\n", port);
+            System.out.printf("\n%d : default server, no offloading route\n", port);
             ServerContext noOffloadsRoute = GrpcServers.forPort(port++)
                     .listenAndAwait(new Greeter.ServiceFactory.Builder().sayHello(noOffloadsStrategy(),
                                     (ctx, request) -> getReplySingle(request, "default server, no offloading route"))
                             .build());
             closeEverything.prepend(noOffloadsRoute);
 
-            // Route attempts no offloads strategy via annotation, which is ignored. (Too late, already offloaded)
+            // Server has default configuration, route attempts to use no offloads strategy via annotation, which is
+            // ignored. (Too late, already offloaded at the server level)
             // -> route offloaded to global executor
-            System.out.printf("\n%d : noOffloadsAnnotation\n", port);
+            System.out.printf("\n%d : default server, no offloading (annotation) route\n", port);
             ServerContext noOffloadsAnnotation = GrpcServers.forPort(port++)
                     .listenAndAwait(new NoOffloadsGreeterService());
             closeEverything.prepend(noOffloadsAnnotation);
 
-            // No offloads strategy specified for all routes by configuration of streaming server.
-            // Route attempts to use no offloads strategy, which is redundant.
+            // No offloads strategy specified on the server, route overrides it to also use no offloads strategy,
+            // which is redundant.
             // -> no offloading, route executed on IoExecutor
-            System.out.printf("\n%d : noOffloadsServerRoute\n", port);
+            System.out.printf("\n%d : no offloading server, no offloading route\n", port);
             ServerContext noOffloadsServerRoute = GrpcServers.forPort(port++)
                     .initializeHttp(init -> init.executionStrategy(noOffloadsStrategy()))
                     .listenAndAwait(new Greeter.ServiceFactory.Builder()
@@ -159,7 +158,7 @@
         return HelloReply.newBuilder().setMessage(reply).build();
     }
 
-    private static class NoOffloadsGreeterService implements Greeter.GreeterService {
+    private static class NoOffloadsGreeterService implements GreeterService {
 
         @Override
         @NoOffloadsRouteExecutionStrategy
```

File: servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcRouter.java [hunk 4]
```diff
@@ -302,7 +311,7 @@ public Completable closeAsync() {
                         public Completable closeAsyncGracefully() {
                             return route.closeAsyncGracefully();
                         }
-                    }, executionStrategy == null ? HttpExecutionStrategies.anyStrategy() : executionStrategy),
+                    }, executionStrategy == null ? HttpExecutionStrategies.defaultStrategy() : executionStrategy),
```
Comment (idelpivnitskiy) [inline servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcRouter.java:314]: Is it expected that here we use `HttpExecutionStrategies.defaultStrategy()`, while line 386 uses `GrpcExecutionStrategies.defaultStrategy()`?

**Review by idelpivnitskiy — APPROVED** [review]
Great work!

**Pull request merged**